framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java

```diff
@@ -139,13 +261,204 @@ default boolean isSubtype(
+   * AnnotationMirror, TypeMirror)} than this method.
+   *
+   * @param subQualifiers set of qualifiers; exactly one per hierarchy
+   * @param superQualifiers set of qualifiers; exactly one per hierarchy
+   * @param typeMirror the type associated with {@code subQualifiers} and {@code superQualifiers}
+   * @return true iff all qualifiers in {@code subQualifiers} are a subqualifier or equal to the
+   *     qualifier in the same hierarchy in {@code superQualifiers}
+   */
+  public boolean isSubtypeShallow(
+      Collection<? extends AnnotationMirror> subQualifiers,
+      Collection<? extends AnnotationMirror> superQualifiers,
+      TypeMirror typeMirror) {
+    return isSubtypeShallow(subQualifiers, typeMirror, superQualifiers, typeMirror);
+  }
+
+  /** A set of annotations and a {@link TypeMirror}. */
+  @AnnotatedFor("nullness")
+  private static class ShallowType {
+
+    /** A set of annotations. */
+    AnnotationMirrorSet annos;
+
+    /** A TypeMirror. */
+    TypeMirror typeMirror;
+
+    /**
+     * Creates a {@code SHallowType}
+     *
+     * @param annos a set of annotations
+     * @param typeMirror a type mirror
+     */
+    private ShallowType(AnnotationMirrorSet annos, TypeMirror typeMirror) {
+      this.annos = annos;
+      this.typeMirror = typeMirror;
+    }
+
+    /**
+     * Creates a {@code ShallowType} where the annotations are the effect annotations on {@code
+     * type} and the type mirror is the underlying type of {@code type}.
+     *
+     * @param type an annotated type to convert to a {@code ShallowType}
+     * @return a shallow type created from {@code type}
+     */
+    @SuppressWarnings("nullness") // AnnotatedTypeMirror isn't annotated for nullness.
+    public static ShallowType create(AnnotatedTypeMirror type) {
+      AnnotatedTypeMirror erasedType = type.getErased();
+      TypeMirror typeMirror =
+          erasedType.getKind() == type.getKind()
+              ? type.getUnderlyingType()
+              : erasedType.getUnderlyingType();
+      return new ShallowType(erasedType.getPrimaryAnnotations(), typeMirror);
+    }
+  }
+
+  /**
+   * Tests whether the effective annotations of {@code subtype} are equal to or are sub-qualifiers
+   * of the effective annotations of {@code supertype}, according to the type qualifier hierarchy.
+   * Other annotations in {@code subtype} and {@code supertype} are ignored.
+   *
+   * <p>The underlying types of {@code subtype} and {@code supertype} are not necessarily in a Java
+   * subtyping relationship with one another and are only used by this method for special cases when
+   * qualifier subtyping depends on the Java basetype.
+   *
+   * @param subtype possible subtype
+   * @param supertype possible supertype
+   * @return true iff the effective annotations of {@code subtype} are equal to or are
+   *     sub-qualifiers of the effective annotations of {@code supertype}
+   */
+  public boolean isSubtypeShallowEffective(
```

Comment:
I'm wondering if the `isSubtypeShallowEffective` methods would make more sense in the `TypeHierarchy` class?  Also, maybe there's a better name?